File: python/array-string/roman-to-integer.py
romans = {'I': 1, 'V': 5, 'X': 10, 'L': 50, 'C': 100, 'D': 500, 'M': 1000}


def roman_to_int(s: str) -> int:
    total = 0
    pointer = len(s) - 1
    while pointer >= 0:
        if pointer > 0 and romans[s[pointer]] > romans[s[pointer - 1]]:
            total += romans[s[pointer]] - romans[s[pointer - 1]]
            pointer -= 2
        else:
            total += romans[s[pointer]]
            pointer -= 1
    return total


def roman_to_int_v2(s: str) -> int:
    result = 0
    for i in range(len(s)):
        if i < len(s) - 1 and romans[s[i]] < romans[s[i + 1]]:
            result -= romans[s[i]]
        else:
            result += romans[s[i]]


    return result


if __name__ == '__main__':
    # s1 = 'III'
    # print(roman_to_int(s1))  # Output: 3
    #
    # s2 = 'IV'
    # print(roman_to_int(s2))  # Output: 4
    #
    # s3 = 'IX'
    # print(roman_to_int(s3))  # Output: 9
    #
    # s4 = 'LVIII'
    # print(roman_to_int(s4))  # Output: 58

    s5 = 'MCMXCIV'
    print(roman_to_int_v2(s5))  # Output: 1994

File: golang/leetcode/array-string/roman-to-integer.go
package main

import "fmt"

var ROMAN2INT = map[string]int{
	"I": 1,
	"V": 5,
	"X": 10,
	"L": 50,
	"C": 100,
	"D": 500,
	"M": 1000,
}

func romanToInt(s string) int {

	pointer := len(s) - 1
	result := 0

	for pointer >= 0 {
		if pointer > 0 && ROMAN2INT[string(s[pointer])] > ROMAN2INT[string(s[pointer-1])] {
			result += ROMAN2INT[string(s[pointer])] - ROMAN2INT[string(s[pointer-1])]
			pointer -= 2
		} else {
			result += ROMAN2INT[string(s[pointer])]
			pointer--
		}
	}
	return result
}

func romanToInt2(s string) int {
	result := 0

	for i := range s {
		if i < len(s)-1 && ROMAN2INT[string(s[i])] < ROMAN2INT[string(s[i+1])] {
			result -= ROMAN2INT[string(s[i])]
		} else {
			result += ROMAN2INT[string(s[i])]
		}
	}
	return result
}

func main() {
	output := romanToInt("III")
	fmt.Println(output) // Output: 3

	output = romanToInt("IV")
	fmt.Println(output) // Output: 4

	output = romanToInt("IX")
	fmt.Println(output) // Output: 9

	output = romanToInt("LVIII")
	fmt.Println(output) // Output: 58

	output = romanToInt2("MCMXCIV")
	fmt.Println(output) // Output: 1994
}
